Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs

```diff
@@ -204,13 +225,29 @@ private Task<IConnection> AcquireAsync()
                 {
                     ThrowObjectDisposedException();
                 }
-                IPooledConnection connection;
 
-                if (!_availableConnections.TryDequeue(out connection))
+                IPooledConnection connection = null;
+                if (!_availableConnections.TryTake(out connection))
                 {
-                    connection = await CreateNewPooledConnectionAsync().ConfigureAwait(false);
+                    Stopwatch connAcquisitionTimer = new Stopwatch();
+                    connAcquisitionTimer.Start();
+                    do
+                    {
+                        if (_poolSize < _maxPoolSize)
+                        {
+                            connection = await CreateNewPooledConnectionAsync().ConfigureAwait(false);
+                            break;
+                        }
+                        // The pool is full at this moment
+                        if (_availableConnections.TryTake(out connection, TimeSpan.FromSeconds(5)))
+                        {
+                            break;
+                        }
+                    } while (connAcquisitionTimer.ElapsedMilliseconds < _connAcquisitionTimeout.TotalMilliseconds);
```

Comment:
Does the connection acquisition timeout applies to individual connection creation or the whole AcquireAsync process? We can decide this later on with the connection timeout thing on sockets.